MLAPI/Serialization/BitWriter.cs

```diff
@@ -182,7 +182,19 @@ public void WriteObjectPacked(object value)
             }
             else if (value.GetType().IsEnum)
             {
-                WriteInt32Packed((int)value);
+                var underlyingType = System.Enum.GetUnderlyingType(value.GetType());
+
+                if (underlyingType == typeof(Int32))
```

Comment:
Is it possible to use a switch case here?

Reply:
Does it even matter? Is the argument for a switch case just to make it easier to read? At the end of the day it just gets converted into the same if statement.

Reply:
Switch cases can be converted to lookup tables after ~5 elements. In improves time complexity from o(n) to o(1), I can confirm that. This improves performance slightly.

Reply:
@MFatihMAR?

Reply:
@Taenebron clearly the compiler is not translating this into a lookup table, are you saying that it should be one instead? Because this code would not be turned into a lookup statement at all by the compiler, and even if the jit optimized for .net backend, IL2CPP will just end up using if/else if/else statements for the switch.

Reply:
I dont really understand the issue here; can you show with a profiler that a switch makes some serious difference? Especially since the types you can inherit from are very limited. 

I'm more concerned with calling ` System.Enum.GetUnderlyingType` every single time.

Reply:
@mattwalsh-unity I already created MTT-507 for this and also #388 would be fixed by a change like this as @TJHeuvel pointed out.

RPC ILPP already does this internally when you pass an `enum` but also adding support to `BitReader`/`BitWriter` with `XXXObjectPacked()` might be worthwhile too.

I'll assign this to myself as well and address it when we prioritize this ticket in a sprint.

Reply:
Also to @wackoisgod's and @Taenebron's point, C# .NET _**does**_ optimize long switch blocks by generating hashmap (Dictionary) lookup behind the scenes but it's not visible and not entirely deterministic from the developer's perspective (unless you double-check IL by yourself all the time). If Unity IL2CPP and/or Unity-modified Mono JIT does something different than the latest C# .NET then that's a different story but we could try to optimize while we can, especially for low-hanging-fruits like this IMHO :)

Reply:
The optimization is at the compilers discretion and it's not deterministic. I wrote an example that never did it with 10 case blocks (and I looked at the IL) so I would argue it's not really an optimization you could even rely on in this use case. 

Also I would argue that the optimization would only be valuable if the area of code was a hot path as the Hash lookup would not be worth it otherwise.

Reply:
Thanks for notes! Interesting.
When I was testing speeds of switch and if-conditions with types I noticed slightly improvement in timings quite early and it was looking stable. Maybe Unity environment / version affects it somehow as suggested.

Reply:
I dont really understand why we're talking about a micro optimalisation like a switch statement, especially given the context. This statement is in a block with 13 other statements above it, if there is concerns about if-vs-switch performance then certainly they will be found there, not in this addition. 

This checks all the possible underlying types of an enum, which is a very limited set (4 types atm), and is hardly ever going to be increased (only when there is a new numeric type introduced, which is never?). I already ordered the cases in such a way that what i think the most common case is (enum inherits int, the default) is first, the second case is byte which will be chosen by performance-concious users. Because of this prioritising, and knowing the data, it might even be faster than a hash lookup. 

My performance concerns are mostly with the GetUnderlyingType call, and the value.GetType call. Additionally when done often the way types are checked can have an impact (https://twitter.com/badamczewski01/status/1361591486740123653). That being said, a lot could potentially have impact, but without profiling we'll never know. Please backup any concerns with proof with a profiler.

Reply:
I mostly refereed to the long process of type detection for incoming object generally. With small amount of cases switch/if is purely aesthetic choice, so I doubt that for that section (4 types) it matters that much. However, for the whole type detection it could be very significant change.

`
switch (obj) //object to write of unknown type
 {
    case int integer:
        var temp1 = integer;
        break;
    case uint uinteger:
        var temp2 = uinteger;
        break;
    case byte b:
        var temp3 = b;
        break;
} 
`